test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.test.ample;
+
+import static org.apache.accumulo.core.client.ConditionalWriter.Status.UNKNOWN;
+import static org.apache.accumulo.core.util.LazySingletons.RANDOM;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.accumulo.core.client.ConditionalWriter;
+import org.apache.accumulo.core.data.ConditionalMutation;
+import org.apache.accumulo.test.ample.metadata.ConditionalWriterInterceptor;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlakyInterceptor implements ConditionalWriterInterceptor {
```

Comment:
This interceptor is a neat way to test the different possibilities. 

The current code is fine so this could be a follow on or just done if we actually need it in the future, I am wondering if we should we make the probabilities configurable? 

I was thinking it could be useful to add another constructor with config parameters to define what percentage of time to pick each action. It could be useful if you want to more heavily test something, or even if you just wanted the interceptor to only perform one action 100% of the time. 

In some cases having things be more deterministic could be useful but that might also be better written as another interceptor if we want deterministic testing since the goal of the Flaky testing is to be random outcomes.

Reply:
More determinism would be nice.  Seems like it would lead to more test, which is probably ok.  Like would have ComprehensiveFlakyAmple1IT that runs w/ pattern A, ComprehensiveFlakyAmple2IT that runs w/ pattern B for the FlakyInterceptor, etc.   Seems like some memory would be needed as multiple rows may be written w/ different collumns, so need reset/restart the pattern when that changes.

If a test fails the randomness may make it hard to debug.  It would probably be generally useful to add a bit more logging around handing of unknown as these should be rare, so it be nice if the code logged some breadcrumbs when they happen and it reacts to them.  That should be a quick change, will add a PR for that.

Reply:
Something else I am trying to figure out w/ this is coverage.  Want to know what code for handling unknown is not being exercised and add test to cover that.

Reply:
Yeah, the flaky test is nice to introduce randomness like a real system but it definitely makes it hard to know what broke and how to reproduce it. So anything that can be added to help track down the steps that caused the error would be helpful. 

There is one more issue with this that I just realized and that is the CI build is generally configured to re-run tests up to 5 times so handle flaky tests. So because of that, if we do get a test failure that is a rare edge case, it is unlikely Jenkins would catch it because it would just re-run and it would pass.

Reply:
> There is one more issue with this that I just realized and that is the CI build is generally configured to re-run tests up to 5 times to handle flaky tests. So because of that, if we do get a test failure that is a rare edge case, it is unlikely Jenkins would catch it because it would just re-run and it would pass.

Yeah that is a problem.  Should we open a follow on issue to explore adding more deterministic tests?

Reply:
This probably needs more discussion but I think the best approach might be to set up the surefire/failsure plugins differently. We really shouldn't need to run most tests with the `rerunFailingTestsCount` flag set as that just hides issues. I think a better approach would be to set up multiple profiles or executions for the test plugins and by default include all tests and don't set the  `rerunFailingTestsCount` flag. Then if we come across flaky tests, we could either try and fix them or if we can't we can `emporarily or permanently add them to another profile where we can set that flag (maybe we need our own -D property for the config). 

In general, the setting is certainly useful because sometimes tests are just hard to write and are flaky so it can be nice to not fail, but it seems like it should be a temporary solution until the test is fixed and not permanent. Keeping it on it definitely can be a problem as it will also hide true errors that need to be fixed but are intermittent.

@ctubbsii - Any thoughts on this and the best way we could set up our Maven test plugins with two different modes/profiles where we only execute a small list of tests with `rerunFailingTestsCount` flag if it's determined we need to? Or do you have another idea?

Reply:
The two different modes that you're talking about are typically covered by a different test suite. Random behavioral tests have historically been covered by something like RandomWalk or ContinuousIngest, which moved to the accumulo-testing repository.

I don't think our integration tests should engage in behavior that results in a random outcome. If there's a random element in the test to cover an edge case, then the test itself should be written such that the edge case is guaranteed, or nearly guaranteed, to have coverage every time.

Example:

```java
@Test
public void randomTestBad() {
  coverEdgeCase20PercentOfTheTime();
}

@Test
public void randomTestBetter() {
  for (int x = 1; x < 100; x++) {
    coverEdgeCase20PercentOfTheTime();
  }
}
```

We can make the Maven build much more complex by having separate executions of different classes of tests, but I would recommend against doing that, because it can get way too complicated way too fast, and I think what we should be doing is writing more resilient ITs that have more comprehensive coverage of these edge cases.